README
alzheimer 0.9

This kernel module erases all the available 
physical memory (RAM) when loaded.
This of course leads to machine "crash" (actually infinite loop).
Use ipmi to restart your server after module is loaded.

Purpose of this module is to prevent physical attacks against RAM. 
Find some events (such as connectivity loss,
cover open, ...) and hook this module on them.
When such event occurs, RAM is immediately erased.
Assuming that you are using full disk encryption
with strong cryptographic algorithm, there 
should be (almost;) no way to get your data.

Tested on i386 and x86_64, feel free to port it on 
different architecture. XXX Multiprocesor systems may not work XXX

If you want to test if it really works,
easiest way is probably using qemu + gdb.
 
Under GNU GPL.

Bug reports and patches welcome.

2010, niekt0@hysteria.sk

-----
Usage:

$ make

To delete memory:

# insmod alzheimer.ko
(good bye;)

You will probably want to hook deleting to some event.
Script hook.sh is generic hook to connectivity loss,
modify it as you need.

n.


ChangeLog
(28.5.2010)

0.5.0 -- First version

0.9.0 -- First public release
	 i386 architecture is working for now.
	 systems with more than 4GB of memory "should" work
1.0.0 -- Added x86_64 support
	 hook.sh modified
	 various fixes


hook.sh
#!/bin/bash

# Under GNU GPL
# 2010 niekt0@hysteria.sk

# modify this script to set exact 
# conditions for pressing red_button

MPATH="."	# path to alzheimer
ALARM_TIME=5	# in "loops"
# how long must be server disconnected, before red button is pushed
CHECK_DELAY=6	# check every..
SERVERS=4	# number of servers
TO_IGNORE=0	# how many servers to ignore even if alive
let TRASHOLD=SERVERS-TO_IGNORE-1;	# how many server can die


# verify_ssl(server)
# certificate should be already in servers/
# to get server certificate, you can use for example:
# echo "" |openssl s_client -connect server:443 2>/dev/null | sed -n '/-----BEGIN CERTIFICATE-----/,/-----END CERTIFICATE-----/p' > servers/server
verify_ssl() {
	local tmp=0;
	
#	echo "verify_ssl $1, $tmp, $fail" >&2;
        echo "" |openssl s_client -connect "$1":443 2>/dev/null >/tmp/hook_$$;
        cat /tmp/hook_$$ | sed -n '/-----BEGIN CERTIFICATE-----/,/-----END CERTIFICATE-----/p' > /tmp/hook_$$_cert;
        cat /tmp/hook_$$ | grep '^SSL handshake has read ' >/dev/null && tmp=1;
        cmp /tmp/hook_$$_cert servers/"$1" >/dev/null || tmp=0;
#	echo "verify_ssl2 $1, $tmp, $fail" >&2;

        if [ "$tmp" == 0 ]; then
                let fail=fail+1;
        fi
	rm -f /tmp/hook_"$$"*

}


# this function performs check itself
check_link() {

# I recommend checking several "friendly" servers,
# located at different places, with big fault tolerance
# Use ssh/ssl for checking (when using SSL, don't trust CA !)

# XXX modify me!
# put your servers here

	local fail=0;

	verify_ssl "hysteria.sk";
	verify_ssl "lol";
	verify_ssl "rofl";
	verify_ssl "xixi";

	echo "$fail"	# result

}

red_button () {

	sync;
	sync;

	#umount -a #?

	echo "RED BUTTON!!!"

	#insmod -f "$MPATH"/alzheimer.ko

	# XXX remove comments this when everything works XXX !!!

}


#main loop

alarm_round=0;

for((;;)); do
	failed=$(check_link);

	if [ "$failed" -gt 0 ]; then
		echo "$failed server/s failed!";
		# XXX send mail? 
	fi
		
	if [ "$failed" -gt "$TRASHOLD" ]; then

		logger -s "Alzheimer: Signal lost!";

		let alarm_round=alarm_round+1;
		if [ "$alarm_round" -gt "$ALARM_TIME" ]; then
			logger -s "Pushing red button";
			red_button;
		fi
	else
		alarm_round=0;
	fi

	sleep "$CHECK_DELAY";

done



arch/x86_64/eraser.S
# Under GNU GPL
# 2010 niekt0@hysteria.sk 
#
# This code should be copied to some free page,
# together with own page tables

# XXX Cloned from i386, not working !!! XXX


# Registers:
#
# EAX : dynamic
# EBX : not used
# ECX : "del_bytes", offset to delete (in 4-bytes just before stos)
# EDX : "end", end of area that should be deleted
# ESI : "start", start of area that should be deleted
# EDI : "del_from", address, where deleting should start in single page
#

eraser_start:
.globl eraser_start
delall:

mov (%rbp),%rsi	# start of area
add $8,%rbp
mov (%rbp),%rdx	# end of area
add $8,%rbp

endlessloop:
cmp $0,%rdx		# if this is last area (end in 0)
je endlessloop		# just loop

delarea:		# delete one area

# exp: map 1 page
# XXX 64
mov %rsi,%rax		# address to delete
mov $0xFFFFFFFFFF000, %rbx 	# hack, cannot and directly
and %rbx,%rax		# only bites 51-12(so this address is 4KB aligned)
add $0x01F,%rax 	# special bits
mov %rax,0x0020	# write to PGD. XXX where is it mapped?
# we will write to linear address 0x4000, =>

invlpg 0x4000	# invalidate "rewrite" page 
	
delpage:

mov $0x4000,%rdi	# address to rewrite, change is done by mapping 
mov $0x200,%rcx		# 4KB to delete, divided by 8 (stosq) 0x1000 / 8 = 200

# just delete page now
# using stosq should be faster than using normal loop
# rewrite rcx "8-bytes" starting from rdi with value of rax

mov $0xADDEADDEADDEADDE,%rax	# rewrite with DEADDEADDEADDEAD ;)
#XXX shr $3,%rcx		# stosl erases 8 bytes per step, so divide by 8
rep stosq		# loop itself

add $0x1000, %rsi	# proceed next page
cmp %rdx,%rsi		# end of area? 
jl delarea		# end > start, so erase next page

jmp delall		# proceed next area

eraser_end:
.globl eraser_end


lkm.c
/* 
 *  This module tries to erase as much RAM as possible, 
 *  before killing itself.
 *  Machine will be left in undefined state, 
 *  but you should be able to reset it over IPMI.
 *  
 *  Under GNU GPL
 *
 *  2010, niekt0@hysteria.sk
 */

#include <linux/init.h>
#include <linux/module.h>
#include <linux/syscalls.h>
#include <linux/kallsyms.h>
#include <linux/moduleparam.h>
#include <linux/mm.h>
#include <linux/vmalloc.h>
#include <linux/raw.h>
#include <linux/ptrace.h>
#include <linux/device.h>
#include <linux/highmem.h>
#include <linux/bootmem.h>
#include <linux/splice.h>
#include <linux/pfn.h>
#include <linux/smp_lock.h>
#include <linux/reboot.h>
#include <asm/reboot.h>

#include "debug.h"

#ifdef CONFIG_IA64
# include <linux/efi.h>
#endif

#define KB_4		4096	

#ifdef ARCH_x86_64
#define NEST_PAGES 	5	// how big is the nest (page tables + asm) x86_64
#else
#define NEST_PAGES 	3	// how big is the nest (page tables + asm)
#endif

#define	MAX_DEL_AREAS	32	// how big is field with areas to delete. ( start+end, => div 2 )a
#define JMPADDR 	0x2000	// static address where to jump

#ifdef ARCH_i386	// 4MB mapping used
#define ARGADDR		(NEST_PAGES*KB_4-MAX_DEL_AREAS*sizeof(unsigned long))	// in new page tables
#endif

#ifdef ARCH_x86_64	// 4KB mapping used
#define ARGADDR         (JMPADDR+KB_4-MAX_DEL_AREAS*sizeof(unsigned long))   // in new page tables
#endif


#ifdef ARCH_i386
#define MY_PAGE_SHIFT	12	// on i386 we need to shift addresses 12, to be able to access over 4GB
#else
#define MY_PAGE_SHIFT	0	// on other architectures this is not needed
#endif


MODULE_LICENSE("GPL");

// Char we show before each debug print
const char program_name[] = "alzheimer";


/* This function is re-implemented, 
 * because older kernels does not have one.
 * 
 * Add "System RAM" type entries from iomem_resource 
 * to our array.
 *
 */
static int my_find_next_system_ram(struct resource *res,unsigned long *memareas,unsigned long nest_physical)
{
        resource_size_t start, end;
        struct resource *p;
	int tmp1=0;

        BUG_ON(!res);

        start = res->start;
        end = res->end;
        BUG_ON(start >= end);

//        read_lock(&resource_lock); //problem on some kernels. We are dirty.
        for (p = iomem_resource.child; p ; p = p->sibling) {
                /* system ram is just marked as IORESOURCE_MEM */
//		dbgprint("%X-%X : %s",p->start,p->end,p->name);
		if ((strncmp("System RAM",p->name,10))==0){
			dbgprint("adding %lX-%lX : %s, [%d]",(unsigned long)p->start,(unsigned long)p->end,p->name,tmp1);
			// XXX TODO test on PAE system with enough RAM
		
			// on i386, all addresses are shifted 12 to the right ("divided")
			// this is needed. if we want to delete addresses over 4GB
			// on i386 architecture. (for example when PAE is used) 
			// so our "units" are 4KB blocks
			memareas[tmp1++]=(p->start >>MY_PAGE_SHIFT); 
			// Exception, for our own code
			if ((p->start <= nest_physical) && \
			    (p->end >= nest_physical)) {
				dbgprint("Splitting area 0x%lX-0x%lX",(unsigned long)p->start,(unsigned long)p->end);
				memareas[tmp1++]=((nest_physical-1) >>MY_PAGE_SHIFT);	//end of first area
				memareas[tmp1++]=((nest_physical+KB_4*NEST_PAGES) >>MY_PAGE_SHIFT);
				// beginning of second area (minus NEST_PAGES pages)
				// this need to be changed if post-jump area is expanded
			}
			memareas[tmp1++]=(p->end >> MY_PAGE_SHIFT);
		}

                if (p->start > end) {
                        p = NULL;
                        break;
                }
        }

//        read_unlock(&resource_lock);
        if (!p)
                return -1;
        /* copy data */
        if (res->start < p->start)
                res->start = p->start;
        if (res->end > p->end)
                res->end = p->end;
        return 0;
}

/*
 * Main function of whole program.
 * Everything is prepared here,
 * and at the end program jump 
 * to minimal assembler routine 
 * with own page tables.
 *
 * - disable all the interrupts,
 * - create fake page tables
 * - copy asm routine close to tables
 * - copy RAM ranges close to asm routine
 * - set new tables and jump to asm routine
 *
 */

int start_cleaning(void){
	
	int tmp2=0;
	unsigned long tmp_long=0;
	unsigned long nest=0;
	unsigned long nest_physical=0;
	unsigned long memareas[MAX_DEL_AREAS];
	extern void eraser_start(void);
	extern void eraser_end(void);

	dbgprint("Cleaning:..");
	
	//Just to make it nicer when debugging
	for (tmp2=0;tmp2<MAX_DEL_AREAS;){
		memareas[tmp2++]=0;
	}


	// Allocate NEST_PAGES pages somewhere,
	// where we will survive the mess
	nest=__get_free_pages(GFP_KERNEL,NEST_PAGES); //XXX 64
	// See include/linux/gfp.h in kernel for details

	nest_physical=virt_to_phys((void *)nest);
	//so we can set up page tables properly


	// Kill all other CPUs
	// This part is ripped from x86_shutdown

	// This function is problematic
	// In new kernels seems to be exported,
	// but there are still some problems. (#ifdef ?)
	// need to be tested on multicpu system

//	nmi_shootdown_cpus((void *)dummy_fc);

	// get memory maps (same as /proc/iomem)
	my_find_next_system_ram(&iomem_resource,memareas,nest_physical);

	dbgprint("nest: 0x%lX",nest);
	dbgprint("Memareas:");
	for (tmp2=0;tmp2<MAX_DEL_AREAS;){
		dbgprint(" 0x%lX",memareas[tmp2++]);
	}

	// fake page_table structure (i386)
	// 4KB non-PAE non-PSE mode is used (2 levels of tables)
	// mixed with 4MB non-PAE PSE mode (only 1 level)
	// We ourselves will be in 4KB page (because of alignment)
	// Working page will be in 4MB PSE mode (so we can access over 4GB)
	// see 24593.pdf from AMD for more info
	
	memset((void *)nest,0,KB_4*NEST_PAGES); //4 KB, NEST_PAGES pages
	
	// copy PGD from original system PGD.
	// This is better, because everything important will stay mapped.
	tmp_long=(unsigned long) phys_to_virt(read_cr3());
	memcpy((void *)nest,(void *)tmp_long,KB_4);


	// Construct own page tables

#ifdef ARCH_i386
	//((nest address + 1 4KB page)=PTE phys addr) + bits
	tmp_long=((nest_physical+KB_4)&0xFFFFF000)+0xF7F;
	memcpy((void *)nest,&tmp_long,sizeof(tmp_long));	// to make 0x0 PGD itself (PGD)
	tmp_long=(nest_physical&0xFFFFF000)+0x3F;		// nest address + bits
	memcpy((void *)nest+KB_4,&tmp_long,sizeof(tmp_long));	// to make 0x0 PGD itself (PT)
	tmp_long=((nest_physical+KB_4*(NEST_PAGES-1))&0xFFFFF000)+0x3F; //nest address + 2 pages + bits (ie. the page with code itself)
	memcpy((void *)nest+KB_4+2*sizeof(unsigned long),&tmp_long,sizeof(tmp_long)); //to map 0x2000 to page with code + ranges

	memcpy((void *)nest+KB_4*(NEST_PAGES-1),eraser_start,eraser_end-eraser_start);// erasing code itself
	memcpy((void *)nest+KB_4*NEST_PAGES-(sizeof(unsigned long)*MAX_DEL_AREAS),memareas,sizeof(unsigned long)*MAX_DEL_AREAS);
	// up to MAX_DEL_AREAS areas is enough for everyone;)
	
	// so code should be placed at 0x2000 (2*4K)
	// arguments should be at 0x2ea0 (3*4K - 64*4)

	// To prevent EIP from containing invalid pointer after table switch (this would cause triple-fault)
	// So EIP must be valid in new tables too.
	// 4MB page used
#endif	
#ifdef ARCH_x86_64
	// ((nest address + 1 4KB page)=PTE phys addr) + bits
	tmp_long=((nest_physical+KB_4)&0xFFFFFFFFFF000)+0x01F;	// bits 51-12 as address, +flag bits (see doc)
	memcpy((void *)nest,&tmp_long,sizeof(tmp_long));	// to make virt 0x0 PTE itself (PML4)
	tmp_long=((nest_physical+KB_4*2)&0xFFFFFFFFFF000)+0x01F;// nest address + bits
	memcpy((void *)nest+KB_4,&tmp_long,sizeof(tmp_long));	// to make virt 0x0 PTE itself (PDPE)
	tmp_long=((nest_physical+KB_4*3)&0xFFFFFFFFFF000)+0x01F;// nest address + bits
	memcpy((void *)nest+KB_4*2,&tmp_long,sizeof(tmp_long));	// to make virt 0x0 PTE itself (PDE)
	tmp_long=((nest_physical+KB_4*3)&0xFFFFFFFFFF000)+0x01F;// nest address + bits
	memcpy((void *)nest+KB_4*3,&tmp_long,sizeof(tmp_long));	// to make 0x0 PTE itself (PTE)

//	tmp_long=((nest_physical)&0xFFFFFFFFFF000)+0x01F;	// nest address + bits
//	memcpy((void *)nest+KB_4*3+sizeof(unsigned long),&tmp_long,sizeof(tmp_long));	// to make 0x1000 PTE itself (PTE)



	tmp_long=((nest_physical+KB_4*(NEST_PAGES-1))&0xFFFFFFFFFF000)+0x01F;
	// nest address + pages + bits 
	// (ie. the page with code itself)
	memcpy((void *)nest+KB_4*3+2*sizeof(unsigned long),&tmp_long,sizeof(tmp_long)); 
	// to map 0x2000 to page 
	// with code + ranges (PTE, rest of tables is recycled)

	memcpy((void *)nest+KB_4*(NEST_PAGES-1),eraser_start,eraser_end-eraser_start);// erasing code itself
	memcpy((void *)nest+KB_4*NEST_PAGES-(sizeof(unsigned long)*MAX_DEL_AREAS),memareas,sizeof(unsigned long)*MAX_DEL_AREAS);
	// up to MAX_DEL_AREAS areas is enough for everyone;)
	
	// so code should be placed at 0x4000 ((5-1)*4K)
	// arguments should be at 0x4ea0 (5*4K - 32*8)

	// To prevent EIP from containing invalid pointer after table switch (this would cause triple-fault)
	// So EIP must be valid in new tables too.
#endif	

	dbgprint("Arguments at: 0x%lX ",(unsigned long)ARGADDR);

	dbgprint("after memset: nest: 0x%lX, nest_phys: 0x%lx",nest,nest_physical);
	
	// disable interrupts 
	__asm ("cli\n\t");

	//Everything is prepared.
	//Just set register and jump now.
	//Good luck!

#ifdef ARCH_i386
	__asm (		
	"mov %%cr4,%%edx\n\t"		// cannot write directly
	"or $0x00000010,%%edx\n\t"      // enable PSE, bit number 4 (because wee need 4MB page)
	"mov %%edx,%%cr4\n\t"		// cannot write directly
	"mov %%eax,%%edx\n\t"		// tmp, stack will be destroyed later. (see last line)
	"mov %%edx,%%cr3\n\t"		// new page table address
	"mov %%cr4,%%edx\n\t"		// cannot write directly
	"and $0xFFFFFFDF,%%edx\n\t"	// disable PAE, bit number 5
	"mov %%edx,%%cr4\n\t"		// cannot write directly
	"mov $0x2000,%%edx\n\t"		// cannot jump directly 
	"invlpg 0x2000\n\t"		// flush TLB 
	"mov %0,%%ebp\n\t"		// 0x2ea0 = 3*4KB-0x160 in our new tables. Good bye stack!
	"jmp *%%edx\n\t"		// JUMP!
	:
	:"i"(ARGADDR),"a"(nest_physical)	// %0 + eax 
	);

#endif

#ifdef ARCH_x86_64

	__asm (		// XXX 64
	"mov %%rax,%%rdx\n\t"		// tmp, stack will be destroyed later. (see last line)
	"mov $0xFFFFFFFFFF000, %%rbx\n\t"	// hack, cannot and directly
	"and %%rbx,%%rdx\n\t"		// only bites 51-12(so this address is 4KB aligned)
	"mov %%rdx,%%cr3\n\t"		// new page table address
	"mov $0x2000,%%rdx\n\t"		// cannot jump directly 
	"invlpg 0x2000\n\t"		// flush TLB 
	"mov %0,%%rbp\n\t"		// Set argument address. Good bye stack!
	"jmp *%%rdx\n\t"		// JUMP!
	:
	:"i"(ARGADDR),"a"(nest_physical)	// %0 + rax 
	);

#endif

#ifdef ARCH_arm

//XXX TODO arm

#endif

	// This point should not be reached
	return 0;

}


/// Function which is executed upon loading this module to kernel
int __init
init_module (void)
{

	dbgprint("init");	

	//no one is going to survive here
	start_cleaning();  

	return 0;
}

/// Function which will be executed upon unloading this module to kernel		
void __exit
cleanup_module (void)
{
	// This should never happen
	// (unless we survived own death;)
	dbgprint ("WTF? Unloading...");

}
